### PR TITLE
A pkgdown website for data.table

### LINKED ISSUE #3675: data.table homepage
(Context: I have [offered](https://twitter.com/hadleywickham/status/1144920847293067264) to help data.table get set up with a basic pkgdown website and provide some advice on the homepage. Matt [suggested](https://twitter.com/MattDowle/status/1145846390796324865) that this was the best place to discuss the homepage)

Package websites in the tidyverse follow a pretty standard format, and I would suggest that data.table mimics this structure, unless there are any strong objections. We usually organise the page as follows:

* Overview: 1-3 paragraph overview of what the package does, often linking to vignettes for more details about particular features. The goal is to help people quickly figure out if they are in the right place.

* Installation: a couple of lines of code so that people can install the package by copying and pasting code (i.e. as easily as possible).

* Cheatsheet: If a cheatsheet is available, high-resolution thumbnails linking to the cheatsheet. This provides some visual interest to the page, and provides a quick overview of the features.

* Usage: 1-2 screens worth of runnable code showing the most important features of the package. 

* Links in the sidebar and navbar to get more details.

(Typically we also echo this material in the readme, so that it's available directly from GitHub.)

Are there any strong feelings about this overall structure?

---

Much of this content can be drawn from the existing data.table wiki, but there are a couple of pieces that where the data.table community needs to guide me:

* I think the first paragraph could be a bit punchier, focussing more on the features of data.table, and providing a little less social proof. Perhaps some of the headline features could be bought up from the bottom of the page, and there could be a new second paragraph that focussed on the impact that data.table has had in the wider community?

* I would suggest combining the basic syntax guide (i.e. the two large images) with links to a cheatsheet. Is there a data.table cheatsheet endorsed by the community? I see that we link to one from [our cheatsheet page](https://github.com/rstudio/cheatsheets/raw/master/datatable.pdf) but I'm not sure if this is official.

* To me, the code samples are a little long. The main goal of the homepage (in my opinion) is to get people interested in the package — the details can go else, but here you want to demonstrate the key features of the package. I think it's also quite important the output is shown (particularly since data.table enhances the print method) and I'd recommend using a dataset that at least evokes a possible analysis context.

### PR DESCRIPTION
**Pull request by hadley**
**jangorecki**: WIP status in https://github.com/Rdatatable/data.table/pull/3677#issuecomment-525239391

----

You can see (a manual deploy) of the site at https://nervous-easley-6de120.netlify.com

Note that to build the pkgdown site yourself you'll need the latest version from GitHub (I had to fix a couple of bugs) and build with `pkgdown::build_site(devel = FALSE, new_process = FALSE)`

This is a very preliminary version presented only in order to have something concrete to respond to. To do:

* [ ] If desired, pick a different theme from https://bootswatch.com
* [ ] Produce a minimal image of the data.table syntax summary images
* [ ] Polish overview paragraphs
* [ ] Review structure of reference page
* [ ] Prepare usage example 

Note that I had to make a few minor changes to data.table sources to work cleanly with pkgdown. I think once you're happy with the basic website, we can discuss in more detail.

Fixes #3675

### CODE REVIEW

File: man/set2key.Rd [hunk 1]
```diff
@@ -4,6 +4,7 @@
 \alias{key2}
 \alias{key<-}
 \title{ Deprecated. }
+\keyword{internal}
```
Comment (MichaelChirico) [inline man/set2key.Rd]: this is for `pkgdown` to skip this in the Reference index, right?

**Comment by MichaelChirico** [comment]
Thanks so much @hadley!

Starting to look things over...

I'm afraid I don't follow you here:

> Produce a minimal image of the data.table syntax summary images

Is that something on your end or for us?

---

I notice the mouse-over of `1.12.3` says `Released version`, but this is the `dev` version (`1.12.2` is on CRAN; we use odd-even versions to distinguish), is that easy to change? Then, on the right, it says `Dev status` but we have a CRAN badge, so the two things are a bit at odds.

---

Am I right that the Changelog page looks for a `foo()` pattern and cross-references the package index to link to help pages? Would it be easy to match `%foo%` as well? In item 8 `between()` links but `%between%` doesn't.

```
between() and %between% are faster for POSIXct, #3519, and now support the .() alias, #2315. Thanks to @Henrik-P for the reports. There is now also support for bit64’s integer64 class and more robust coercion of types, #3517.
```

---

For the `_pkgdown.yml` config, we can use any of the `\alias{}`es to refer to a page right? And all of the non-`\keyword{internal}` pages must be indexed? Is it OK to comma-separate multiple of a pages aliases (e.g. I would switch `IDateTime` to `IDate, ITime, IDateTime`)? Is it OK to have duplicate references to a ref page? It might make sense to have `IDate, ITime, IDateTime`, then `year, month, hour, day` separately so that all the functions _do_ get a callout on the reference page (even though they're all in the same Rd)

---

In general if a fix/change/upgrade is required in `pkgdown`, would be happy to try and help with the caveat that some primer about how things are working would be helpful.

File: man/set2key.Rd (hunk 1 above)
Comment (hadley) [reply man/set2key.Rd]: Yeah, plus it excludes it from the regular package index (which I think is good practice).

**Comment by hadley** [comment]
To be precise, I think someone on the data table needs to redraw the [minimal syntax image](https://github.com/Rdatatable/data.table/wiki/pictures/syntax1.jpg) to remove the background colour, remove pdf download link in top right, and refine (or remove) the title.

---

Can you file an issue in pkgdown for this? Currently we assume that dev versions look like 1.0.0.9000, so we'll need to think about how to parameterise that differently.

---

I don't remember the exact details of how the autolinking works, but I think it should be straightforward to also autolink infix operators. Can you please file an issue?

---

There's currently very little control over how the aliases are display on the reference page. So far my position has been that if you think there are too many, then you should consider breaking up the topic into multiple pieces.

---

I don't have pkgdown front of mind at the moment, so I don't have any general advice, sorry 😞

**Comment by MichaelChirico** [comment]
Re: minimal syntax image -- I had a go at reproducing that in text in my commit. Will leave it up to Matt how to proceed.

Re: issues: [filed](https://github.com/r-lib/pkgdown/issues/1081) and [filed](https://github.com/r-lib/pkgdown/issues/1082)

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -303,6 +305,10 @@ integration: # merging all artifacts to produce single R repository and summarie
     - Rscript -e 'pdf.copy("data.table", "test-rel-lin")'
     # web/checks/check_results_$pkg.html
     - Rscript -e 'check.index("data.table", names(test.jobs))'
+    # pkgdown
+    - apt-get update -qq && apt-get install -y libxml2-dev
+    - Rscript -e 'install.packages("remotes", repos=Sys.getenv("CRAN_MIRROR"), quiet=TRUE); remotes::install_github("r-lib/pkgdown", repos=Sys.getenv("CRAN_MIRROR"), quiet=TRUE); install.packages("data.table", repos=file.path("file:",normalizePath("bus/integration/cran")), quiet=TRUE); pkgdown::build_site(devel=FALSE, new_process=TRUE, override=list(destination="./pkgdown"))'
```
Comment (jangorecki) [inline .gitlab-ci.yml]: We need to speed it up, currently it installs `remotes`, `pkgdown` and 59! other packages. Plus it requires OS dependency libxml2-dev. As a result this job takes 16 minutes while previously took 5 minutes.

File: README.md [hunk 3]
```diff
@@ -2,56 +2,81 @@
...
+
+### Cheatsheets
 
 <a href="https://github.com/rstudio/cheatsheets/raw/master/datatable.pdf"><img src="man/figures/cheatsheet.png" width="615" height="242"/></a>
 
-## Example
+## Community
 
-This is a basic example which shows you how to solve a common problem:
+`data.table` is widely used by the R community. As of July 2019, it was used by over 680 CRAN and Bioconductor packages and was the [9th most starred](http://www.r-pkg.org/starred) R package on GitHub. If you need help, the `data.table` community is active StackOverflow, with nearly [9,000 questions](http://stackoverflow.com/questions/tagged/data.table).
```
[the remark this answers is not in the document]
Comment (jangorecki) [reply README.md]: AFAIR rendering README.Rmd to md requires `devel=TRUE`, which tries to use roxygen on your package. When we don't use roxygen it raises many warnings, and modify NAMESPACE file.
It also makes it difficult to keep Rmd to md in sync for a front page on github project. On every change made to Rmd developer should render it and include md in the commit too. And that would introduce inconsistencies due to minor changes in render engines. Different dev might have different rmarkdown/pandoc version. Then better to skip md in the repo at all.
As we don't really need any dynamically computed content there I don't see the point to use Rmd. This info doesn't have to the most up to day, updating once a year is more than enough.

**Comment by jangorecki** [comment]
Thanks @hadley.
I tweaked your PR, I hope it is getting closer to to-be state.
For now I am publishing it at https://jangorecki.gitlab.io/data.table/
Still on my list:
- well categorise topics in References
- investigate why Changelog is empty
- gather feedback
- reduce pkgdown time overhead to our CI
- add google analytics

File: .gitlab-ci.yml (hunk 2 above)
Comment (jangorecki) [reply .gitlab-ci.yml]: GitLab caching is currently in use when mirror suggested dependencies. We do publish all of them together with devel data.table in our R repository, to achieve ultimate reproducibility.
As of now we do not need to use any other package in the pipeline, all CI is in base R. Thus adding 60 packages related to CI only is pretty big thing. We need to make a secondary CI-related cache/repo.
Just for reference, `mirror-packages` job:
https://github.com/Rdatatable/data.table/blob/81af9b6dd7522d91ea86d9e944f6b43a7d21c022/.gitlab-ci.yml#L20-L37

Comment (jangorecki) [reply .gitlab-ci.yml]: after recent commits to pkgdown we went down to 13 minutes for this job
so 3/11 of the overhead is fixed

File: _pkgdown.yml [hunk 4]
```diff
@@ -1,5 +1,8 @@
 url: https://jangorecki.gitlab.io/data.table
 
+development:
+  version_tooltip: "Odd versions are in development; even are on CRAN"
```
Comment (jangorecki) [inline _pkgdown.yml]: we can safely use "development version" as we won't publish even releases in CI, even are cran-only

**Comment by jangorecki** [comment]
Once we agree it is ready to merge I will then update url references from `jangorecki.gitlab.io` to `Rdatatable.gitlab.io` and restrict build to `master` branch only (during dev it also publish for `pkgdown`).

**Comment by MichaelChirico** [comment]
### Changelog

- It's quite nice to have the internal links to old versions. We also have `NEWS.0.md`, is there any way to reference that? It so happens that we did link it at the bottom but that's pretty obscure. It might be nice to have the a toggle at the top, or for the last link on the right to redirect to that file on GH? @hadley this might be a `pkgdown` FR?

- Section headers are also inconsistent @hadley:

Screenshot 2019-08-30 at 10 00 43 AM
Screenshot 2019-08-30 at 10 00 48 AM

For some reason the date as parsed shows up on the left _only for v1.12.2_?

- I see this, feels a bit redundant, anything to be done? `(in development) Unreleased`

- Is there a way to add the dates to the links on the right as well?

### Manual

- `nafill() setnafill()` appear twice, is that intentional? If so I would also add `copy()` to the in-place modification section (and generally double-reference anything that feels kind-of borderline)

### Vignettes

- in Intro vignette, `in the next vignette` at the end -- not totally clear what this is referring to

**Comment by jangorecki** [comment]
> It's quite nice to have the internal links to old versions. We also have NEWS.0.md, is there any way to reference that? It so happens that we did link it at the bottom but that's pretty obscure. It might be nice to have the a toggle at the top, or for the last link on the right to redirect to that file on GH? @hadley this might be a pkgdown FR?

This is already possible, we can make a menu of changelog files, but I don't see it that much important.

> date

related issue: https://github.com/r-lib/pkgdown/issues/1118
maybe worth to put your comments there

File: DESCRIPTION [hunk 5]
```diff
@@ -49,7 +49,7 @@ Suggests: bit64, curl, R.utils, knitr, xts, nanotime, zoo, yaml
 SystemRequirements: zlib
 Description: Fast aggregation of large data (e.g. 100GB in RAM), fast ordered joins, fast add/modify/delete of columns by group using no copies at all, list columns, friendly and fast character-separated-value read/write. Offers a natural and flexible syntax, for faster development.
 License: MPL-2.0 | file LICENSE
-URL: http://r-datatable.com
+URL: https://github.com/Rdatatable/data.table
```
Comment (mattdowle) [inline DESCRIPTION]: Why this change?

Comment (jangorecki) [reply DESCRIPTION]: I am not precisely sure, https://pkgdown.r-lib.org/articles/linking.html vignette mention to use github repo url. I reverted back to use two urls, as initially commited by Hadley

Comment (jangorecki) [reply DESCRIPTION]: I am not sure but we might need to use three urls there, as pkgdown website will be published under: https://Rdatatable.gitlab.io/data.table
